File: packages/core/src/markdown.ts
/**
 * KRO-131 — tokenizador de markdown INLINE (PURO, agnóstico de plataforma).
 *
 * Para fields con `behavior: 'markdown'` (lore/relato/descripcion…). Convierte un
 * string en una lista PLANA de tokens que cada plataforma renderiza con su piel
 * (`@kromia/react` → JSX, `@kromia/flutter` → `TextSpan`), garantizando markdown
 * IDÉNTICO en ambas sin reimplementar el parser en cada lado.
 *
 * Soporta: `**negrita**` · `*cursiva*` / `_cursiva_` · `` `code` `` ·
 * `[texto](url)` · saltos de línea (token `break`). NO es un parser de bloque
 * (sin listas/headings/tablas) — cubre el grueso del lore de cartas sin meter
 * dependencias. Para prosa rica de wiki, Studio usa su `<Markdown>` completo;
 * esto es solo para el render de cartas/secciones.
 */

export type MarkdownToken =
  | { type: 'text';   value: string }
  | { type: 'bold';   value: string }
  | { type: 'italic'; value: string }
  | { type: 'code';   value: string }
  | { type: 'link';   value: string; href: string }
  | { type: 'break' };

/** Tokeniza UNA línea (sin `\n`) a tokens inline. */
function tokenizeLine(text: string): MarkdownToken[] {
  const out: MarkdownToken[] = [];
  // Orden de precedencia: negrita (**) · code (`) · link ([..](..)) · cursiva (* o _).
  const re = /\*\*([^*]+)\*\*|`([^`]+)`|\[([^\]]+)\]\(([^)]+)\)|\*([^*]+)\*|_([^_]+)_/g;
  let last = 0;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    if (m.index > last) out.push({ type: 'text', value: text.slice(last, m.index) });
    if (m[1] != null)      out.push({ type: 'bold',   value: m[1] });
    else if (m[2] != null) out.push({ type: 'code',   value: m[2] });
    else if (m[3] != null) out.push({ type: 'link',   value: m[3], href: m[4] });
    else if (m[5] != null) out.push({ type: 'italic', value: m[5] });
    else if (m[6] != null) out.push({ type: 'italic', value: m[6] });
    last = re.lastIndex;
  }
  if (last < text.length) out.push({ type: 'text', value: text.slice(last) });
  return out;
}

/**
 * Tokeniza un string markdown inline COMPLETO. Los saltos de línea se emiten
 * como tokens `break` entre líneas, así el render de cada plataforma solo recorre
 * la lista (sin volver a partir por `\n`). Texto sin marcas → un único token
 * `text`.
 */
export function parseInlineMarkdown(text: string): MarkdownToken[] {
  const out: MarkdownToken[] = [];
  const lines = (text ?? '').split('\n');
  lines.forEach((line, i) => {
    if (i > 0) out.push({ type: 'break' });
    out.push(...tokenizeLine(line));
  });
  return out;
}

File: packages/core_dart/lib/src/markdown.dart
/// KRO-131 — tokenizador de markdown INLINE (PURO). Espejo 1:1 de `markdown.ts`.
///
/// Para fields `behavior:'markdown'` (lore/relato/descripcion…). Convierte un
/// string en una lista PLANA de tokens que cada plataforma renderiza con su piel
/// (@kromia/react → JSX, @kromia/flutter → TextSpan) → markdown IDÉNTICO sin
/// reimplementar el parser. Soporta `**negrita**` · `*cursiva*`/`_cursiva_` ·
/// `` `code` `` · `[texto](url)` · saltos de línea (token `break`). No es parser
/// de bloque.
library;

/// Token de markdown inline. `type`: `text|bold|italic|code|link|break`.
/// `value` = contenido (vacío en `break`); `href` solo en `link`.
class MarkdownToken {
  final String type;
  final String value;
  final String? href;
  const MarkdownToken(this.type, {this.value = '', this.href});
}

// Precedencia: negrita (**) · code (`) · link ([..](..)) · cursiva (* o _).
final RegExp _inlineRe =
    RegExp(r'\*\*([^*]+)\*\*|`([^`]+)`|\[([^\]]+)\]\(([^)]+)\)|\*([^*]+)\*|_([^_]+)_');

/// Tokeniza UNA línea (sin `\n`) a tokens inline.
List<MarkdownToken> _tokenizeLine(String text) {
  final out = <MarkdownToken>[];
  var last = 0;
  for (final m in _inlineRe.allMatches(text)) {
    if (m.start > last) {
      out.add(MarkdownToken('text', value: text.substring(last, m.start)));
    }
    if (m.group(1) != null) {
      out.add(MarkdownToken('bold', value: m.group(1)!));
    } else if (m.group(2) != null) {
      out.add(MarkdownToken('code', value: m.group(2)!));
    } else if (m.group(3) != null) {
      out.add(MarkdownToken('link', value: m.group(3)!, href: m.group(4)));
    } else if (m.group(5) != null) {
      out.add(MarkdownToken('italic', value: m.group(5)!));
    } else if (m.group(6) != null) {
      out.add(MarkdownToken('italic', value: m.group(6)!));
    }
    last = m.end;
  }
  if (last < text.length) {
    out.add(MarkdownToken('text', value: text.substring(last)));
  }
  return out;
}

/// Tokeniza un string markdown inline completo. Los `\n` → tokens `break` entre
/// líneas, así el render de cada plataforma solo recorre la lista.
List<MarkdownToken> parseInlineMarkdown(String text) {
  final out = <MarkdownToken>[];
  final lines = text.split('\n');
  for (var i = 0; i < lines.length; i++) {
    if (i > 0) out.add(const MarkdownToken('break'));
    out.addAll(_tokenizeLine(lines[i]));
  }
  return out;
}
